sdk_v2/kubeflow/training/api/training_client.py

```diff
@@ -12,7 +12,553 @@
+            RuntimeError: Failed to list TrainJobs.
+        """
+
+        result = []
+        try:
+            thread = self.custom_api.list_namespaced_custom_object(
+                constants.GROUP,
+                constants.VERSION,
+                self.namespace,
+                constants.TRAINJOB_PLURAL,
+                async_req=True,
+            )
+            response = thread.get(constants.DEFAULT_TIMEOUT)
+
+            for item in response["items"]:
+                # If runtime ref is set, we check the TrainJob's runtime.
+                if (
+                    runtime_ref is not None
+                    and item["spec"]["runtimeRef"]["name"] != runtime_ref
+                ):
+                    continue
+                trainjob = self.api_client.deserialize(
+                    utils.FakeResponse(item),
+                    models.KubeflowOrgV2alpha1TrainJob,
+                )
+                result.append(self.__get_trainjob_from_crd(trainjob))  # type: ignore
+
+        except multiprocessing.TimeoutError:
+            raise TimeoutError(
+                f"Timeout to list {constants.TRAINJOB_KIND}s in namespace: {self.namespace}"
+            )
+        except Exception:
+            raise RuntimeError(
+                f"Failed to list {constants.TRAINJOB_KIND}s in namespace: {self.namespace}"
+            )
+
+        return result
+
+    def get_job(self, name: str) -> types.TrainJob:
+        """Get the TrainJob information"""
+
+        try:
+            thread = self.custom_api.get_namespaced_custom_object(
+                constants.GROUP,
+                constants.VERSION,
+                self.namespace,
+                constants.TRAINJOB_PLURAL,
+                name,
+                async_req=True,
+            )
+
+            trainjob = self.api_client.deserialize(
+                utils.FakeResponse(thread.get(constants.DEFAULT_TIMEOUT)),  # type: ignore
+                models.KubeflowOrgV2alpha1TrainJob,
+            )
+
+        except multiprocessing.TimeoutError:
+            raise TimeoutError(
+                f"Timeout to get {constants.TRAINJOB_KIND}: {self.namespace}/{name}"
+            )
+        except Exception:
+            raise RuntimeError(
+                f"Failed to get {constants.TRAINJOB_KIND}: {self.namespace}/{name}"
+            )
+
+        return self.__get_trainjob_from_crd(trainjob)  # type: ignore
+
+    def get_job_logs(
+        self,
+        name: str,
+        follow: bool = False,
+        component: str = constants.JOB_TRAINER_NODE,
+        node_index: int = 0,
+    ) -> Dict[str, str]:
+        """Get the logs from TrainJob
+        TODO (andreyvelich): Should we change node_index to node_rank ?
+        TODO (andreyvelich): For the initializer, we can add the unit argument.
+        """
+
+        pod_name = None
+        # Get Initializer or Trainer Pod name.
+        for c in self.get_job(name).components:
```

Comment:
What if there are multiple Pods for the same role like initializer? How can we identify the Pod name?
We can imagine the situation where the Pods are recreate (but the old one still exist) based on restart, recreate, success, failure policies.

Reply:
Hmm, it is a good point.
I think, in [the current loop ](https://github.com/kubeflow/training-operator/blob/f921bbe5b458c82faf57c3521aa7b6e1d9260e4c/sdk_v2/kubeflow/training/api/training_client.py#L505) it return the following:
```py
Component(name='initializer', status='Failed', device='cpu', device_count='10', pod_name='xe37e16669f7-initializer-0-0-gaf51')
Component(name='initializer', status='Succeeded', device='cpu', device_count='10', pod_name='xe37e16669f7-initializer-0-0-hl4wv')
Component(name='trainer-node-0', status='Succeeded', device='gpu', device_count='4', pod_name='xe37e16669f7-trainer-node-0-0-ktbt8')
```
I think, we should discuss if that experience looks good to our users.

WDYT @tenzen-y ?

Reply:
IIRC, the katib UI had similar problem, and we addressed that, previously.
We might be able to learn something from that.

https://github.com/kubeflow/katib/blob/2b41ae62ab3905984e02123218351a703c03bf56/pkg/ui/v1beta1/backend.go#L720

Reply:
I see. I guess, that function can return logs from the Failed Pod even if we have another Succeeded Pod since the podList order is alphabetical. Is that understanding correct, @tenzen-y ?
What do you think we should return with `get_job()` API in that case ?

Reply:
> I see. I guess, that function can return logs from the Failed Pod even if we have another Succeeded Pod since the podList order is alphabetical. Is that understanding correct, @tenzen-y ?

I guess that this current log collection implementation obtains the logs randomly. So, we sometimes fetch the logs from Failed Pods even if the replaced Running or Succeeded Pods, isn't it?

Reply:
@tenzen-y As you can see right now we fetch logs only from the Trainer component, if `follow=True`: https://github.com/kubeflow/training-operator/blob/e9c22b2b477f4beea5c96006f6997e8212f4ee4c/sdk_v2/kubeflow/training/api/training_client.py#L368-L377


Otherwise, we just return logs from all TrainJob components: https://github.com/kubeflow/training-operator/blob/e9c22b2b477f4beea5c96006f6997e8212f4ee4c/sdk_v2/kubeflow/training/api/training_client.py#L410-L432


Additionally, by default the Job controller will not create additional Pods if it fails. It will try to re-create the existing Pod until it reaches BackOffLimits.


I am happy to revisit this API in the following PR if we want to improve it.
WDYT @tenzen-y ?

Reply:
@Electronic-Waste @akshaychitneni @shravan-achar @droctothorpe @deepanker13 @seanlaii @astefanutti Any thoughts/ideas on the above topic ?

Reply:
> Additionally, by default the Job controller will not create additional Pods if it fails

The behavior of the failure is different based on Pod restart Policy, batch job success, and failure policies. So, we can easily reproduce the unexpected log fetching situations.

I can easily imagine the situations where the v2 SDK users create issues to report unexpected logs seeing.
So, I believe that addressing the issue would be better before the first v2 release.

If you can work on the issue in another PR before we release the first v2 release, I'm ok for now.

Reply:
Sure, created this @tenzen-y: https://github.com/kubeflow/training-operator/issues/2348

Reply:
Thank you!